### ISSUE TITLE
context.markVariableAsUsed does not work consistently

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#9810]
**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** 9.2.1
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
'use strict';

const { Linter } = require('eslint');
const linter = new Linter();

linter.defineRule('test-rule', context => ({
  'Program:exit'() {
    const ast = context.getSourceCode().ast;
    context.getDeclaredVariables(ast.body[0])
      .forEach(variable => context.markVariableAsUsed(variable.name));
  }
}));

const lintResults = linter.verify(
  'function foo() {}',
  {
    rules: {
      'no-unused-vars': 'error',
      'test-rule': 'error'
    },
    parserOptions: {
      sourceType: 'module'
    }
  }
);


console.log(lintResults);
```

```bash
node thatFile.js
```

**What did you expect to happen?**

I expected the list of lint results to be empty, because there is an unused variable in the program which gets marked with `context.markVariableAsUsed`.

**What actually happened? Please include the actual, raw output from ESLint.**

An error is reported:

```
[ { ruleId: 'no-unused-vars',
    severity: 2,
    message: '\'foo\' is defined but never used.',
    line: 1,
    column: 10,
    nodeType: 'Identifier',
    source: 'function foo() {}',
    endLine: 1,
    endColumn: 13 } ]
```

Applying either of the following changes causes the error to disappear:

* Changing the `Program:exit` listener in `test-rule` to a `Program` listener
* Changing the order of `no-unused-vars` and `test-rule` in the config object

This is happening as a result of a design flaw in the `context.markVariableAsUsed` API. When a rule uses `context.markVarkableAsUsed`, it mutates a shared scope analysis object. The `no-unused-vars` rule checks the scope analysis object in a `Program:exit` listener, and avoids reporting an unused variable if the corresponding object has been marked by another rule. The problem is that another rule could call `context.markVarkableAsUsed` after the `Program:exit` listener in `no-unused-vars` has already run.

ESLint does not make any guarantees about rule ordering, because rules are not supposed to be able to communicate with each other. (`context.markVarkableAsUsed` is an exception to this.) Given that we now allow parsers to provide custom scope analysis, I think the best solution might just involve deprecating `context.markVariableAsUsed` in favor of custom scope analysis. However, if we do take that route, we would probably need to provide a way for plugins to provide scope analysis info as well, so that (e.g.) the `react/jsx-uses-vars` rule can be replaced with a scope analysis modification.

**Comment by mysticatea** [comment]
> I think the best solution might just involve deprecating context.markVariableAsUsed in favor of custom scope analysis.

Interesting idea. 
But I don't think that it's possible because we can't apply multiple custom parsers (or scope analysis) to a kind of code. Currently, users can use a mix of multiple plugins which use `context.markVariableAsUsed`, or can use `babel-eslint` (it has own scope-analysis) and `eslint-plugin-react` (it uses `context.markVariableAsUsed`) in the same time.

I think better if we clarify it, we can't use `context.markVariableAsUsed` in `Program:exit` selector, in the doc.

**Comment by j-f1** [comment]
Since `no-unused-vars` is a special case, would it make sense to hardcode it into the linter so it runs after linting is complete?

**Comment by not-an-aardvark** [comment]
@mysticatea I agree, that would probably be simplest. It seems undesirable to have an API that depends on rule interference, but in this case it's probably not worth changing.

One possible solution would be a change to the plugin API like this, but it might not be worthwhile:

```js
module.exports = {
  rules: // ...
  configs: // ...
  variablesToMarkAsUsed(scope) {
    return [
      // ... list of variables
    ];
  };
```

**Comment by not-an-aardvark** [comment]
@j-f1 I would rather not add a special case for `no-unused-vars`, if possible -- it seems like it would set a precedent for adding more features to rules that interfere with the behavior of core rules, which I would like to avoid in the future.

**Comment by platinumazure** [comment]
> One possible solution would be a change to the plugin API like this, but it might not be worthwhile:

That won't work for react/jsx-uses-vars, which interprets the JSX tag names in a file and marks those variables as used.

**Comment by not-an-aardvark** [comment]
Oops, you're right. That API would probably need to provide the AST too.

**Comment by ilyavolodin** [comment]
To the best of my knowledge, this API is not used too heavily. It was mainly introduced to be able allow eslint-plugin-react to correctly mark variables as used. It seems to work for them just fine, so I'm not sure there's a good enough reason to fix this.

**Comment by not-an-aardvark** [comment]
I agree that it might not be worth fixing this, although I think the lack of widespread usage makes a fix *more* compelling, not less (since we could plausibly deprecate and remove it in the future).

My main motivation for wanting to get rid of `context.markVariableAsUsed` is that it is designed around communication between rules, whereas other parts of ESLint are designed around the assumption that rules can't communicate. This leads to strange cases like the one described in this issue, where the communication mechanism is unsound.

**Comment by mysticatea** [comment]
Plugin-level hook is interesting.
The `variablesToMarkAsUsed` is to cancel `no-unused-vars` reports. I think that we can generalize it for all rules.

```js
module.exports.hooks = {
    onReport(event) {
        console.log(event.message) // the reported message.
        console.log(event.context) // the rule context which reported it. `.getScope()`, `.getSourceCode()`, `.text`, and `.ast` are available.
        console.log(event.cancel()) // remove the message.
    }
}
```

**Comment by not-an-aardvark** [comment]
I'm not sure I like the idea of plugins being able to cancel arbitrary reports from rules. It seems like that could cause confusion if a plugin is silencing a message from a rule that the user explicitly enables. I think it would be better to use `eslint-rule-composer` instead.

(Having said that, I suppose plugins can already do this by using a processor.)

**Comment by mysticatea** [comment]
I think, we cannot use `eslint-rule-composer` as a combination of multiple plugins as same as the original proposal of this issue.

**Comment by nzakas** [comment]
@not-an-aardvark you originally opened this issue. Is it something you're planning on continuing to work on?

**Comment by not-an-aardvark** [comment]
No, I'll close this.
